airbyte-commons-worker/src/main/java/io/airbyte/workers/general/DefaultReplicationWorker.java

```diff
@@ -549,6 +575,47 @@ private static void validateSchema(final RecordSchemaValidator recordSchemaValid
     }
   }
 
+  /**
+   * Generates a map from stream -> the explicit list of fields included for that stream, according to
+   * the configured catalog. Since the configured catalog only includes the selected fields, this lets
+   * us filter records to only the fields explicitly requested.
+   *
+   * @param catalog
+   * @param streamToSelectedFields
+   */
+  private static void populatedStreamToSelectedFields(final ConfiguredAirbyteCatalog catalog,
+                                                      final Map<AirbyteStreamNameNamespacePair, List<String>> streamToSelectedFields) {
+    for (final var s : catalog.getStreams()) {
+      final List<String> selectedFields = new ArrayList<>();
+      final JsonNode propertiesNode = s.getStream().getJsonSchema().findPath("properties");
+      if (propertiesNode.isObject()) {
+        propertiesNode.fieldNames().forEachRemaining((fieldName) -> selectedFields.add(fieldName));
+      } else {
+        throw new RuntimeException("No properties node in stream schema");
+      }
+      streamToSelectedFields.put(AirbyteStreamNameNamespacePair.fromConfiguredAirbyteSteam(s), selectedFields);
+    }
+  }
+
+  private static void filterSelectedFields(final Map<AirbyteStreamNameNamespacePair, List<String>> streamToSelectedFields,
+                                           final AirbyteMessage airbyteMessage) {
+    final AirbyteRecordMessage record = airbyteMessage.getRecord();
+
+    if (record == null) {
+      // This isn't a record message, so we don't need to do any filtering.
+      return;
+    }
+
+    final AirbyteStreamNameNamespacePair messageStream = AirbyteStreamNameNamespacePair.fromRecordMessage(record);
+    final List<String> selectedFields = streamToSelectedFields.getOrDefault(messageStream, Collections.emptyList());
+    final JsonNode data = record.getData();
+    if (data.isObject()) {
+      ((ObjectNode) data).retain(selectedFields);
+    } else {
+      throw new RuntimeException("Unexpected data in record");
```

Comment:
Probably a good idea to include the record's body in this error message because if it ever happens, I think we'll want to know what the record looked like!